src/appleseed/renderer/kernel/lighting/pathtracer.h

```diff
@@ -718,25 +726,184 @@ bool PathTracer<PathVisitor, VolumeVisitor, Adjoint>::march(
+        }
+
+        // Continue path tracing if sampled distance exceeds total length of the ray,
+        // otherwise process the scattering event.
+        if (extinction_is_null || volume_ray.m_tmax < distance_sample)
+        {
+            Spectrum transmission;
+            volume->evaluate_transmission(
+                vertex.m_volume_data,
+                volume_ray,
+                transmission);
+
+            vertex.m_throughput *= transmission;
+            vertex.m_throughput /=                       // equivalent to multiplying by MIS weight
+                foundation::average_value(transmission); // and then dividing by transmission[channel]
+
+            break;
+        }
+
+        //
+        // Bounce.
+        //
+
+        // Terminate the path if this scattering event is not accepted.
+        if (!m_volume_visitor.accept_scattering(vertex.m_prev_mode))
+            return false;
+
+        // Let the volume visitor handle the scattering event.
+        m_volume_visitor.on_scatter(vertex);
+
+        // Retrieve scattering spectrum.
+        Spectrum scattering_coef = volume->scattering_coefficient(
+            vertex.m_volume_data, volume_ray);
+        if (scattering_coef.size() != channel_count)
+            Spectrum::upgrade(scattering_coef, scattering_coef);
+
+        // Evaluate transmission between the origin and the sampled distance.
         Spectrum transmission;
-        volume->evaluate_transmission(data, volume_ray, transmission);
+        volume->evaluate_transmission(
+            vertex.m_volume_data,
+            volume_ray,
+            distance_sample,
+            transmission);
+        if (transmission.size() != channel_count)
+            Spectrum::upgrade(transmission, transmission);
+
+        // Upgrade throughput spectrum, if necessary.
+        if (vertex.m_throughput.size() != channel_count)
+            Spectrum::upgrade(vertex.m_throughput, vertex.m_throughput);
+        
+        // Compute MIS weight.
+        // MIS terms are:
+        //  - scattering albedo,
+        //  - throughput of the entire path up to the sampled point.
+        // Reference: "Practical and Controllable Subsurface Scattering
+        // for Production Path Tracing", p. 1 [ACM 2016 Article].
+        float mis_weights_sum = 0.0f;
+        for (size_t i = 0; i < channel_count; ++i)
+        {
+            if (extinction_coef[i] > 1.0e-6f)
+                mis_weights_sum += transmission[i] * scattering_coef[i] / extinction_coef[i];
+        }
+        if (mis_weights_sum < 1.0e-6f)
+        {
+            return false;  // no scattering
+        }
+        const float current_mis_weight =
+            channel_count *
+            transmission[channel] *
+            scattering_coef[channel] /
+            extinction_coef[channel] /
+            mis_weights_sum;
+
+        vertex.m_throughput *= scattering_coef;
         vertex.m_throughput *= transmission;
+        vertex.m_throughput /= distance_pdf;
```

Comment:
You can avoid one division and a bunch of multiplications by grouping the scalar factors together:

    vertex.m_throughput *= current_mis_weight / distance_pdf;